File: leetcode/1137.py
"""
Runtime: 64 ms, faster than 5.04% of Python3 online submissions for N-th Tribonacci Number.
Memory Usage: 14.2 MB, less than 45.35% of Python3 online submissions for N-th Tribonacci Number.
"""
# performance optimization
class Trib:
    def __init__(self):
        self.dp = [0, 1, 1] + [0] * 35
        for i in range(3, 38):
            self.dp[i] = self.dp[i - 1] + self.dp[i - 2] + self.dp[i - 3]


class Solution:
    t = Trib()

    def tribonacci(self, n: int) -> int:
        return self.t.dp[n]


# space optimization
class Solution:
    def tribonacci(self, n: int) -> int:
        a, b, c, d = 0, 1, 1, 2
        if n < 4:
            return [a, b, c, d][n]
        for i in range(n - 3):
            d, c, b, a = d + c + b, d, c, b

        return d

File: leetcode/1137.cpp
// Runtime: 0 ms (beats 100%)
// Memory: 6 MB (beats 36.35%)
// recursion with a template
template <int n> class Tribonacci {
 public:
  constexpr static int res = (Tribonacci<n - 1>::res + Tribonacci<n - 2>::res +
                              Tribonacci<n - 3>::res);
  Tribonacci<n - 1> lowerTrib;
  constexpr int operator[](int i) { return i == n ? res : lowerTrib[i]; }
};

// base cases for 0, 1, 2
template <> class Tribonacci<0> {
 public:
  constexpr static int res = 0;
  constexpr int operator[](int i) { return res; }
};

template <> class Tribonacci<1> {
 public:
  constexpr static int res = 1;
  Tribonacci<0> lowerTrib;
  constexpr int operator[](int i) { return i == 1 ? res : lowerTrib[i]; }
};

template <> class Tribonacci<2> {
 public:
  constexpr static int res = Tribonacci<1>::res + Tribonacci<0>::res;
  Tribonacci<1> lowerTrib;
  constexpr int operator[](int i) { return i == 2 ? res : lowerTrib[i]; }
};

Tribonacci<37> t;

class Solution {
 public:
  int tribonacci(int n) { return t[n]; }
};
